tests/compiler.js

```diff
@@ -485,6 +485,27 @@
                   { name: 'james' },
                   'FooInclude james');
 
+            equal('{% macro foo() %}{% set x = "foo"%}{{ x }}{% endmacro %}' +
+                  '{% set x = "bar" %}' +
+                  '{{ x }}' +
+                  '{{ foo() }}' +
+                  '{{ x }}',
+                  'barfoobar');
+
+            // this test checks that the issue #577 is resolved
+            equal('{% macro foo(topLevel = "false", prefix = "") %}' +
+                  '{% set x = prefix + "foo" %}' +
+                  '{% if topLevel %}' +
+                  '{% for i in [1,2,3] %}' +
+                  '{{ foo(false, x) }}' +
+                  '{% endfor %}' +
+                  '{% else %}' +
+                  '{{ x }}' +
+                  '{% endif %}' +
+                  '{% endmacro %}' +
+                  '{{ foo(true) }}',
+                  'foofoofoofoofoofoo');
```

Comment:
I think this test could really use a brief comment explaining what the failure mode would look like. Just something along these lines: 

> If the bug is not fixed, and set variables leak into the caller scope, there will be too many "foo"s here, because each recursive call will append a "foo" to the variable `x` in its caller's scope, instead of just its own.